chainntnfs/txconfnotifier.go

```diff
@@ -98,78 +101,116 @@ func NewTxConfNotifier(startHeight uint32, reorgSafetyLimit uint32) *TxConfNotif
-		tcn.confNotifications[*ntfn.TxID] =
-			append(tcn.confNotifications[*ntfn.TxID], ntfn)
-		return nil
+	tcn.Lock()
+	defer tcn.Unlock()
+
+	tcn.confNotifications[*ntfn.TxID] = append(
+		tcn.confNotifications[*ntfn.TxID], ntfn,
+	)
+
+	return nil
+}
+
+// UpdateConfDetails attempts to mark an existing unconfirmed transaction as
+// confirmed. The confirmation details will be used to dispatch notifications to
+// the active clients which requested them.
+func (tcn *TxConfNotifier) UpdateConfDetails(txid chainhash.Hash,
+	details *TxConfirmation) error {
+
+	select {
+	case <-tcn.quit:
+		return errors.New("TxConfNotifier is exiting")
+	default:
 	}
 
-	// If the transaction already has the required confirmations, we'll
-	// dispatch the notification immediately.
-	confHeight := txConf.BlockHeight + ntfn.NumConfirmations - 1
-	if confHeight <= tcn.currentHeight {
-		Log.Infof("Dispatching %v conf notification for %v",
-			ntfn.NumConfirmations, ntfn.TxID)
+	tcn.Lock()
+	defer tcn.Unlock()
 
-		// We'll send a 0 value to the Updates channel, indicating that
-		// the transaction has already been confirmed.
-		select {
-		case <-tcn.quit:
-			return fmt.Errorf("TxConfNotifier is exiting")
-		case ntfn.Event.Updates <- 0:
-		}
+	// First, we'll determine whether we have any active notifications for
+	// this transaction.
+	ntfns, ok := tcn.confNotifications[txid]
+	if !ok {
+		return errors.New("no active notification requests found")
+	}
 
-		select {
-		case <-tcn.quit:
-			return fmt.Errorf("TxConfNotifier is exiting")
-		case ntfn.Event.Confirmed <- txConf:
-			ntfn.dispatched = true
-		}
-	} else {
-		// Otherwise, we'll record the transaction along with the height
-		// at which we should notify the client.
-		ntfn.details = txConf
-		ntfnSet, exists := tcn.ntfnsByConfirmHeight[confHeight]
-		if !exists {
-			ntfnSet = make(map[*ConfNtfn]struct{})
-			tcn.ntfnsByConfirmHeight[confHeight] = ntfnSet
+	// Now, for each notification request, we'll attempt to dispatch a
+	// notification to the caller if the transaction has met its required
+	// number of confirmations.
+	for _, ntfn := range ntfns {
+		// Skip the notification request if a notification has already
+		// been dispatched.
+		if ntfn.dispatched {
+			continue
 		}
-		ntfnSet[ntfn] = struct{}{}
 
-		// We'll also send an update to the client of how many
-		// confirmations are left for the transaction to be confirmed.
-		numConfsLeft := confHeight - tcn.currentHeight
-		select {
-		case ntfn.Event.Updates <- numConfsLeft:
-		case <-tcn.quit:
-			return errors.New("TxConfNotifier is exiting")
+		confHeight := details.BlockHeight + ntfn.NumConfirmations - 1
+		if confHeight <= tcn.currentHeight {
+			Log.Infof("Dispatching %v conf notification for %v",
+				ntfn.NumConfirmations, ntfn.TxID)
+
+			// We'll send a 0 value to the Updates channel,
+			// indicating that the transaction has already been
+			// confirmed.
+			select {
+			case ntfn.Event.Updates <- 0:
```

Comment:
Can we risk the channel having a buffer one element to small in this case? (since it is created with `numconfs` capacity)

Reply:
It's safe to do so as we don't send an update indicating the required number of confirmations. Now that I think about this though, it should do that in place of sending an update indicating there aren't any confirmations left as that's already handled by the `Confirmed` channel. Will address it in a follow-up PR.